libs/ojsonnet/Eval_jsonnet.ml

```diff
@@ -102,63 +76,73 @@ let log_call env str tk =
-  | V.Object _ -> "object"
-  | V.Array _ -> "array"
-  | V.Function _ -> "function"
-
-let std_primivite_equals _env (v : V.value_) (v' : V.value_) : bool =
+  | Primitive (Null _) -> "null"
+  | Primitive (Bool _) -> "boolean"
+  | Primitive (Double _) -> "number"
+  | Primitive (Str _) -> "string"
+  | ObjectVal _ -> "object"
+  | Array _ -> "array"
+  | Function _ -> "function"
+
+let std_primivite_equals _env (v : value_) (v' : value_) : bool =
   match (v, v') with
-  | V.Primitive p, V.Primitive p' -> (
+  | Primitive p, Primitive p' -> (
       match (p, p') with
       (* alt: use deriving and Primitive.eq *)
-      | V.Null _, V.Null _ -> true
-      | V.Bool (b, _), V.Bool (b', _) -> b =:= b'
-      | V.Str (s, _), V.Str (s', _) -> s = s'
-      | V.Double (f, _), V.Double (f', _) -> f =*= f'
-      | V.Null _, _
-      | V.Bool _, _
-      | V.Str _, _
-      | V.Double _, _ ->
+      | Null _, Null _ -> true
+      | Bool (b, _), Bool (b', _) -> b =:= b'
+      | Str (s, _), Str (s', _) -> s = s'
+      | Double (f, _), Double (f', _) -> f =*= f'
+      | Null _, _
+      | Bool _, _
+      | Str _, _
+      | Double _, _ ->
           false)
   (* Should we raise an exn if one of the value is not a primitive?
    * No, the spec seems to not restrict what v and v' can be.
    *)
   | _else_ -> false
 
+let rec lookup env tk local_id =
+  let entry =
+    try Map_.find local_id env.locals with
+    | Not_found ->
+        error tk
+          (spf "could not find '%s' in the environment"
+             (string_of_local_id local_id))
+  in
+  match entry.value with
+  | Val v -> v
+  | Unevaluated e -> eval_expr entry.env e
 (*****************************************************************************)
 (* eval_expr *)
 (*****************************************************************************)
 
-let rec eval_expr (env : env) (v : expr) : V.value_ =
+and eval_expr (env : env) (v : expr) : value_ =
   match v with
   | L v ->
       let prim =
         match v with
-        | A.Null tk -> V.Null tk
-        | A.Bool (b, tk) -> V.Bool (b, tk)
-        | A.Str x -> V.Str (A.string_of_string_ x)
+        | A.Null tk -> Null tk
+        | A.Bool (b, tk) -> Bool (b, tk)
+        | A.Str x -> Str (A.string_of_string_ x)
         | A.Number (s, tk) ->
             (* TODO: double check things *)
             let f = float_of_string s in
-            V.Double (f, tk)
+            Double (f, tk)
       in
-      V.Primitive prim
+      Primitive prim
   (* lazy evaluation of Array elements and Functions *)
   | Array (l, xs, r) ->
       let elts =
-        xs |> Array.of_list
-        |> Array.map (fun e ->
-               let v = lazy (eval_expr env e) in
-               { V.v; e })
+        xs
+        |> Common.map (fun x -> { value = Unevaluated x; env })
+        |> Array.of_list
       in
-      V.Array (l, elts, r)
-  | Lambda v -> V.Function v
+      Array (l, elts, r)
+  | Lambda v -> Function v
```

Comment:
Again, better keep Lambda, so it mimics more the parallel between Core_jsonnet and Value_jsonnet.